auth/src/providers/VSCodeAzureSubscriptionProvider.ts

```diff
@@ -0,0 +1,191 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import type { AzureAccount } from '../contracts/AzureAccount';
+import type { AzureSubscription, SubscriptionId, TenantId } from '../contracts/AzureSubscription';
+import type { GetOptions, GetSubscriptionsOptions, TenantIdAndAccount } from '../contracts/AzureSubscriptionProvider';
+import type { AzureTenant } from '../contracts/AzureTenant';
+import { dedupeSubscriptions } from '../utils/dedupeSubscriptions';
+import { NotSignedInError } from '../utils/NotSignedInError';
+import { AzureSubscriptionProviderBase, DefaultGetOptions, DefaultGetSubscriptionsOptions } from './AzureSubscriptionProviderBase';
+
+/**
+ * Extension of {@link AzureSubscriptionProviderBase} that adds caching of accounts, tenants, and subscriptions,
+ * as well as filtering and deduplication according to configured settings.
+ */
+export class VSCodeAzureSubscriptionProvider extends AzureSubscriptionProviderBase {
+    /**
+     * Cache of accounts.
+     */
+    private readonly accountCache: Set<AzureAccount> = new Set();
+
+    /**
+     * Cache of tenants. The key is the account ID, lowercase.
+     */
+    private readonly tenantCache: Map<string, AzureTenant[]> = new Map();
+
+    /**
+     * Cache of subscriptions. The key is `${accountId}/${tenantId}`, lowercase.
+     */
+    private readonly subscriptionCache: Map<string, AzureSubscription[]> = new Map();
+
+    /**
+     * @inheritdoc
+     */
+    public override async getAccounts(options: GetOptions = DefaultGetOptions): Promise<AzureAccount[]> {
+        if (options.noCache) {
+            this.accountCache.clear();
+        }
+
+        // If needed, refill the cache
+        if (this.accountCache.size === 0) {
+            const accounts = await super.getAccounts(options);
+            if (accounts.length === 0) {
+                this.logger?.debug('auth: No accounts found');
+                throw new NotSignedInError();
```

Comment:
Special logic to throw `NotSignedInError` if no accounts are listed at all. Should we move it up to the base class?